File: src/main/java/de/johanneswirth/apps/authservice/services/PasswordChangeService.java
package de.johanneswirth.apps.authservice.services;

import com.codahale.metrics.annotation.ExceptionMetered;
import com.codahale.metrics.annotation.Timed;
import de.johanneswirth.apps.authservice.AuthDAO;
import de.johanneswirth.apps.authservice.PasswordUtils;
import de.johanneswirth.apps.common.IStatus;
import de.johanneswirth.apps.common.Secured;
import org.jdbi.v3.core.Jdbi;

import javax.validation.Valid;
import javax.validation.constraints.NotNull;
import javax.ws.rs.*;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.SecurityContext;

import static de.johanneswirth.apps.common.SuccessStatus.OK;

@Path("change-pass")
public class PasswordChangeService {

    private AuthDAO dao;
    private PasswordUtils utils;

    public PasswordChangeService(Jdbi jdbi, PasswordUtils utils) {
        this.dao = jdbi.onDemand(AuthDAO.class);
        this.utils = utils;
    }

    @POST
    @Consumes({MediaType.APPLICATION_JSON})
    @Produces({MediaType.APPLICATION_JSON})
    @Secured
    @Valid
    @NotNull
    @Timed
    @ExceptionMetered
    public IStatus register(@NotNull @Valid Password password, @Context SecurityContext securityContext) {
        String salt = utils.getSalt(30);
        String pass = utils.generateSecurePassword(password.password, salt);
        dao.setPassword(pass, salt, Integer.parseInt(securityContext.getUserPrincipal().getName()));
        return OK;
    }

    @OPTIONS
    public void options() {}

    private class Password {
        public String password;

        public Password() { }
    }
}

File: src/main/java/de/johanneswirth/apps/auth/services/PasswordChangeService.scala
package de.johanneswirth.apps.auth.services

import com.codahale.metrics.annotation.{ExceptionMetered, Timed}
import de.johanneswirth.apps.auth.PasswordUtils
import de.johanneswirth.apps.auth.jdbi.AuthDAO
import de.johanneswirth.apps.common.{IStatus, Secured}
import de.johanneswirth.apps.common.SuccessStatus.OK
import org.jdbi.v3.core.Jdbi

import javax.validation.Valid
import javax.validation.constraints.NotNull
import javax.ws.rs._
import javax.ws.rs.core.{Context, MediaType, SecurityContext}

@Path("change-pass")
class PasswordChangeService(val jdbi: Jdbi, var utils: PasswordUtils) {
  private val dao = jdbi.onDemand(classOf[AuthDAO])

  @POST
  @Consumes(Array(Array(MediaType.APPLICATION_JSON)))
  @Produces(Array(Array(MediaType.APPLICATION_JSON)))
  @Secured
  @Valid
  @NotNull
  @Timed
  @ExceptionMetered
  def register(@NotNull @Valid password: PasswordChangeService#Password, @Context securityContext: SecurityContext): IStatus = {
    val salt = utils.getSalt(30)
    val pass = utils.generateSecurePassword(password.password, salt)
    dao.setPassword(pass, salt, securityContext.getUserPrincipal.getName.toInt)
    OK
  }

  @OPTIONS def options() = {
  }

  private case class Password(password: String)
}